### PR TITLE
GSB: Propagate ArchetypeResolutionKind to resolveDependentMemberTypes

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
…when resolving equivalence classes

<s>Fixes infinite recursion in a fixed compiler crasher.</s>

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
@slavapestov I am hoping this won't perturb your work on the GSB

@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility release

File: lib/AST/GenericSignatureBuilder.cpp [hunk 1]
```diff
@@ -3554,7 +3555,8 @@ static Type resolveDependentMemberTypes(GenericSignatureBuilder &builder,
         equivClass->recursiveConcreteType = false;
```
Comment (slavapestov) [inline lib/AST/GenericSignatureBuilder.cpp:3555]: Somewhat unrelated: Should we be setting this in maybeResolveEquivalenceClass() as well?

**Review by slavapestov — APPROVED** [review]

File: lib/AST/GenericSignatureBuilder.cpp (hunk 1 above)
Comment (AnthonyLatsis) [reply lib/AST/GenericSignatureBuilder.cpp:3555]: I was actually wrong, there is no infinite recursion on master. This is cherry-picked from a branch that relaxes recursive superclass constraints, which happens to be the cause of the recursion this is supposed to fix, but I misunderstood how `not` works and running into the same problem with the 5.1.3 release convinced me the infinite recursion is present on master as well, although I didn't even bother to check.

So, answering your question – probably yes, but it seems we will not have something to test against until after we lift the restriction on these recursive constraints.

Still doesn't hurt having this though.

Comment (slavapestov) [reply lib/AST/GenericSignatureBuilder.cpp:3555]: I would prefer it if the crashers in validation-tests did not check for precise diagnostic messages, and instead asserted that the compiler rejected the code without crashing (which is what `not swift -frontend -typecheck` accomplishes)

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please clean smoke test OS X

**Pull request merged**